src/HOCON/Impl/HoconObject.cs

```diff
@@ -29,223 +31,315 @@ namespace Hocon
+        /// <param name="result">When this method returns, contains the <see cref="HoconField"/> 
+        /// associated with the specified key, if the key is found; 
+        /// otherwise, null. This parameter is passed uninitialized.</param>
+        /// <returns>
+        /// <c>true</c> if the <see cref="HoconObject"/> contains a field with the the specified key; otherwise, <c>false</c>.
+        /// </returns>
+        public bool TryGetField(string key, out HoconField result)
         {
-            FloatingObjects.Add(this);
-            Items = new Dictionary<string, HoconValue>();
+            return TryGetValue(key, out result);
         }
 
-        /// <summary>
-        /// Retrieves the underlying map that contains the barebones
-        /// object values.
-        /// </summary>
-        public IDictionary<string, object> Unwrapped
+        public HoconField GetField(HoconPath path)
         {
-            get
+            if (path == null)
+                throw new ArgumentNullException(nameof(path));
+
+            if (path.Count == 0)
+                throw new ArgumentException("Path is empty.", nameof(path));
+
+            var pathIndex = 0;
+            var currentObject = this;
+            while (true)
             {
-                return Items.ToDictionary(k => k.Key, v =>
-                {
-                    HoconObject obj = v.Value.GetObject();
-                    if (obj != null)
-                        return (object) obj.Unwrapped;
-                    return v.Value;
-                });
+                var key = path[pathIndex];
+
+                if (!currentObject.TryGetValue(key, out var field))
+                    throw new KeyNotFoundException($"Could not find field with key `{key}` at path `{new HoconPath(path.GetRange(0, pathIndex + 1)).Value}`");
+
+                if (pathIndex >= path.Count - 1)
+                    return field;
+
+                if (field.Type != HoconType.Object)
+                    throw new HoconException($"Invalid path, trying to access a key on a non object field. Path: `{new HoconPath(path.GetRange(0, pathIndex + 1)).Value}`");
+
+                currentObject = field.GetObject();
+                pathIndex = pathIndex + 1;
             }
         }
 
-        /// <summary>
-        /// Retrieves the underlying map that this element is based on.
-        /// </summary>
-        public Dictionary<string, HoconValue> Items { get; }
-
-        /// <summary>
-        /// Determines whether this element is a string.
-        /// </summary>
-        /// <returns><c>false</c></returns>
-        public bool IsString()
+        public bool TryGetField(HoconPath path, out HoconField result)
         {
-            return false;
+            result = null;
+            try
+            {
+                result = GetField(path);
+            }
+            catch
+            {
+                return false;
```

Comment:
Remember that throwing exceptions is thousands times more expensive than just returning a boolean. Keeping that in mind I think you should implement `TryGetField` as a default, and then use it to implement `GetField`. Not other way around.

Reply:
Won't I loose the fine grain control for throwing the exact exception for each error types that way? I was trying to kill 2 birds with 1 stone by also throwing meaningful user friendly exception messages for the error.